crates/ruff_linter/src/rules/refurb/rules/slice_to_remove_prefix_or_suffix.rs

```diff
@@ -0,0 +1,479 @@
+    test: &'a ast::Expr,
+    else_or_target: &'a ast::Expr,
+    slice: &'a ast::Expr,
+) -> Option<RemoveAffixData<'a>> {
+    let compr_value = ast::comparable::ComparableExpr::from(value);
+    let compr_else_or_target = ast::comparable::ComparableExpr::from(else_or_target);
+    if compr_value != compr_else_or_target {
+        return None;
+    }
+    let slice = slice.as_slice_expr()?;
+    let compr_test_expr = ast::comparable::ComparableExpr::from(
+        &test.as_call_expr()?.func.as_attribute_expr()?.value,
+    );
+    let func_name = test
+        .as_call_expr()?
+        .func
+        .as_attribute_expr()?
+        .attr
+        .id
+        .as_str();
+
+    let func_args = &test.as_call_expr()?.arguments.args;
+
+    let [affix] = func_args.as_ref() else {
+        return None;
+    };
+    if compr_value != compr_test_expr || compr_test_expr != compr_else_or_target {
+        return None;
+    }
+    let (affix_kind, bound) = match func_name {
+        "startswith" if slice.upper.is_none() => (AffixKind::StartsWith, slice.lower.as_ref()?),
+        "endswith" if slice.lower.is_none() => (AffixKind::EndsWith, slice.upper.as_ref()?),
+        _ => return None,
+    };
+    Some(RemoveAffixData {
+        text: value,
+        bound,
+        affix_query: AffixQuery {
+            kind: affix_kind,
+            affix,
+        },
+    })
+}
+
+/// Tests whether the slice of the given string actually removes the
+/// detected affix.
+///
+/// For example, in the situation
+///
+/// ```python
+///  text[:bound] if text.endswith(suffix) else text
+/// ```
+///
+/// This function verifies that `bound == -len(suffix)` in two cases:
+///   - `suffix` is a string literal and `bound` is a number literal
+///   - `suffix` is an expression and `bound` is
+///     exactly `-len(suffix)` (as AST nodes, prior to evaluation.)
+fn affix_matches_slice_bound(data: &RemoveAffixData, semantic: &SemanticModel) -> bool {
+    let RemoveAffixData {
+        text: _,
+        bound,
+        affix_query: AffixQuery { kind, affix },
+    } = *data;
+
+    match (kind, bound, affix) {
+        (
+            AffixKind::StartsWith,
+            ast::Expr::NumberLiteral(ast::ExprNumberLiteral {
+                value: num,
+                range: _,
+            }),
+            ast::Expr::StringLiteral(ast::ExprStringLiteral {
+                range: _,
+                value: string_val,
+            }),
+        ) => num.as_int().is_some_and(|x| {
+            // Only support prefix removal for size at most `u32::MAX`
+            u32::try_from(string_val.len()).is_ok_and(|length| x == &ast::Int::from(length))
+        }),
+        (
+            AffixKind::StartsWith,
+            ast::Expr::Call(ast::ExprCall {
+                range: _,
+                func,
+                arguments,
+            }),
+            _,
+        ) => {
+            semantic.match_builtin_expr(func, "len")
+                && arguments.len() == 1
```

Comment:
`semantic.match_builtin_expr()` is unfortunately more expensive than what you had before (we need to resolve the qualified name, which involves several hashmap lookups and some complex AST matching), so we should probably switch the order of these conditions:

```suggestion
            arguments.len() == 1
                && semantic.match_builtin_expr(func, "len")
```